pkg/process/monitor/process_monitor.go

```diff
@@ -0,0 +1,247 @@
+)
+
+type ProcessMonitor struct {
+	m  sync.Mutex
+	wg sync.WaitGroup
+
+	state int
+
+	events chan netlink.ProcEvent
+	done   chan struct{}
+	errors chan error
+
+	procEventCallbacks map[ProcessEventType][]*ProcessCallback
+	callbackRunner     chan func()
+	callbackRunnerDone chan struct{}
+}
+
+type ProcessEventType int
+
+const (
+	EXEC ProcessEventType = iota
+	FORK
+	EXIT
+)
+
+type ProcessMetadataField int
+
+const (
+	ANY ProcessMetadataField = iota
+	NAME
+	MAPFILE
+)
+
+type ProcessCallback struct {
+	Event    ProcessEventType
+	Metadata ProcessMetadataField
+	Regex    *regexp.Regexp
+	Callback func(pid uint32)
+}
+
+func (pm *ProcessMonitor) enqueueCallback(callback *ProcessCallback, pid uint32) {
+	pm.callbackRunner <- func() { callback.Callback(pid) }
+}
+
+func (pm *ProcessMonitor) InitWithAllCurrentProcess() {
+	fn := func(pid int) error {
+		pm.m.Lock()
+		for _, c := range pm.procEventCallbacks[EXEC] {
+			pm.evalEXECCallback(c, uint32(pid))
+		}
+		pm.m.Unlock()
+		return nil
+	}
+
+	if err := util.WithAllProcs(util.HostProc(), fn); err != nil {
+		log.Errorf("process monitor init, scanning all process failed %s", err)
+	}
+
+	pm.state = RUNNING
+}
+
+func (p *ProcessMonitor) evalEXECCallback(c *ProcessCallback, pid uint32) {
+	if c.Metadata == ANY {
+		p.enqueueCallback(c, pid)
+		return
+	}
+
+	proc, err := process.NewProcess(int32(pid))
+	if err != nil {
+		log.Errorf("process %d parsing failed %s", pid, err)
+		return
+	}
+	switch c.Metadata {
+	case NAME:
+		pname, err := proc.Name()
+		if err != nil {
+			log.Errorf("process %d name parsing failed %s", pid, err)
+			return
+		}
+		if c.Regex.MatchString(pname) {
+			p.enqueueCallback(c, pid)
+		}
+	case MAPFILE:
+		mmaps, err := proc.MemoryMaps(true)
+		if err != nil {
+			log.Errorf("process %d/smaps parsing failed %s", pid, err)
+			return
+		}
+		if mmaps == nil {
+			return
+		}
+		for _, mmap := range *mmaps {
+			if c.Regex.MatchString(mmap.Path) {
+				p.enqueueCallback(c, pid)
+			}
+		}
+	}
+}
+
+func (pm *ProcessMonitor) Subscribe(callback *ProcessCallback) (UnSubscribe func()) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	pm.procEventCallbacks[callback.Event] = append(pm.procEventCallbacks[callback.Event], callback)
+
+	return func() {
+		pm.m.Lock()
+		defer pm.m.Unlock()
+
+		for i, c := range pm.procEventCallbacks[callback.Event] {
+			if c == callback {
+				l := len(pm.procEventCallbacks[callback.Event])
+				pm.procEventCallbacks[callback.Event][i] = pm.procEventCallbacks[callback.Event][l-1]
+				pm.procEventCallbacks[callback.Event] = pm.procEventCallbacks[callback.Event][:l-1]
+			}
+		}
+	}
+}
+
+func (pm *ProcessMonitor) Stop() {
+	close(pm.callbackRunnerDone)
+	close(pm.done)
+	pm.wg.Wait()
+	processMonitor = nil
+}
+
+func GetProcessMonitor(cfg *config.Config) *ProcessMonitor {
+	// we want only one instance as
+	// netlink.ProcEventMonitor() netlink doesn't support multiple connections
+	if processMonitor != nil {
+		return processMonitor
+	}
```

Comment:
that's not thread safe, and thus that does not guarantee a single instance
you can achieve single instance by one of the following methods:

initializing `processMonitor` in `init` method
```go
func init() {
        processMonitor = newProcessMonitor()
}
```
another method, to init during the declaration
```go
var (
        processMonitor = newProcessMonitor()
)
```

another option would be to define a global mutex `mu` and to be used in  `GetProcessMonitor()`

```go
var (
        processMonitor *ProcessMonitor
        mu sync.Mutex
)
...

func GetProcessMonitor(cfg *config.Config) *ProcessMonitor {
        mu.Lock()
        defer mu.Unlock()
	// we want only one instance as
	// netlink.ProcEventMonitor() netlink doesn't support multiple connections
	if processMonitor != nil {
		return processMonitor
	}
        ...
}
```

and another option would be using `sync.Once`

```go
func GetProcessMonitor(cfg *config.Config) *ProcessMonitor {
        sync.Once(initProcessMonitor)
        return processMonitor
}
```

Reply:
Yes the init is not thread safe at the initialization/cleanup, but there are only one path to the init/cleanup.

Reply:
note: config would be need at the initialisation time